allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -13,15 +16,37 @@ class _LazyInstances(Iterable):
     An ``Iterable`` that just wraps a thunk for generating instances and calls it for
     each call to ``__iter__``.
     """
-    def __init__(self, instance_generator: Callable[[], Iterator[Instance]]) -> None:
+    def __init__(self,
+                 instance_generator: Callable[[], Iterable[Instance]],
+                 cache_file: str = None,
+                 deserialize: Callable[[str], Instance] = None,
+                 serialize: Callable[[Instance], str] = None) -> None:
         super().__init__()
         self.instance_generator = instance_generator
+        self.cache_file = cache_file
+        self.deserialize = deserialize
+        self.serialize = serialize
 
     def __iter__(self) -> Iterator[Instance]:
-        instances = self.instance_generator()
-        if isinstance(instances, list):
-            raise ConfigurationError("For a lazy dataset reader, _read() must return a generator")
-        return instances
+        # Case 1: Use cached instances
+        if self.cache_file is not None and os.path.exists(self.cache_file):
+            with open(self.cache_file) as f:
+                for line in f:
+                    yield self.deserialize(line)
+        # Case 2: Need to cache instances
+        elif self.cache_file is not None:
+            with open(self.cache_file, 'w') as f:
+                for instance in self.instance_generator():
+                    f.write(self.serialize(instance))
+                    f.write("\n")
+                    yield instance
+        # Case 3: No cache
+        else:
+            instances = self.instance_generator()
+            if isinstance(instances, list):
```

Comment:
now that we're using LazyInstances in both cases, this check is no longer correct (and is causing tests to fail).

I'm not sure how to fix it

Reply:
Hmm, I thought I checked the tests...  Looks like it's a couple I didn't run locally.  I'll figure it out.

Reply:
I believe this is because your `LazyFakeReader` class isn't actually lazy, even though it says it is: https://github.com/allenai/allennlp/blob/9dec020281ee9521e7f1ffd696bcbb102c399703/allennlp/tests/commands/train_test.py#L173-L184

So this new logic appears to be more correct than the old logic, because this test should have failed earlier.

Similarly, the DROP reader test is marked as being both lazy and non-lazy, but the reader did not `yield`, it returned a list.  I'm not sure why the previous logic didn't crash in this case.